Documented some performance tuning lessons. (#485)
digital-asset/docs.daml.com: docs/2.8.0/docs/canton/usermanual/performance.rst

If you choose lower resource limits, you may observe a lower latency, at the cost of lower throughput and
commands getting rejected with the error code ``PARTICIPANT_BACKPRESSURE``.

.. _size_of_connection_pools:

**Size of connection pools.** Make sure that every node uses a connection pool to communicate with the database.
This avoids the extra cost of creating a new connection on every database query.

> Canton chooses a suitable connection pool by default, but for performance sensitive applications you may want to optimise.

Author: fixed

on a single participant node. A bad model will reach a fraction of that. Therefore, it is important to understand the connection between the model 
and the performance implications. This section aims to give a few guidelines.

.. _model_tuning_views:

Reduce the Number of Views
~~~~~~~~~~~~~~~~~~~~~~~~~~
One key performance driver in Canton is the transaction structure resulting from the model. A Daml command is effectively 
a "program" that computes a transaction structure. This transaction structure is then broken up by Canton into 
pieces called "transaction views". The transaction views are then encrypted and sent to the participants who then confirm
each view. Each view requires cryptography and creates additional payload that needs to be processed and validated. 

As a result, the performance of your system directly depends on how many views the transaction creates. The number of views 
of a transaction is logged on the participant side as a DEBUG log message: 

.. code:: 
   
    Computed transaction tree with total=27 for #root-nodes=8

> If you submit a command (which can have multiple Daml commands), then every Daml command will create a new so-called "root-node",
> which in term creates a new view each time. 

Reviewer: Technically, the LAPI command "createAndExercise" creates TWO root nodes.
```suggestion
If you submit a ledger API command (which can have multiple Daml commands), then every Daml command will create one so-called root nodes (two for the ``CreateAndExercise`` command).
Each root in turn creates one view. 
```

Just putting all Daml commands into a single batch command will not help, as you might still create lots of views.

In the protocol up to version 5, Canton will create a view for every action node in the transaction tree if the informees change 
between the node and its parent. Therefore, in order to minimise the number of views, you should try to minimise the number of 

> changes of informees. Informees are the signatories, observers of the contract and the stakeholders of the choice. 

Reviewer: I don't think we have defined the concept of stakeholders of a choice anywhere.
```suggestion
changes of informees. Informees are roughly the signatories, observers of the contract and of the choice, and the controllers of the choice.
```

> Instead of sending a batch with 20 exercise commands `Foo`, group the contracts you have by stakeholder and send one exercise 

Reviewer: ```suggestion
Instead of sending a batch with 20 exercise commands `Foo`, group your contracts by stakeholder and send one exercise 
```

> command on a generator contract with the list of contracts that in turn exercises the `Foo` choice per stakeholder group.

Reviewer: Would it make sense to give an example Daml code for this transformation here?

The rule of thumb is: the number of views should depend on the number of stakeholder groups you have in your batch choice, 
and be independent of the number of "batches" you process in parallel within one command.

Reduce Ledger Events 
~~~~~~~~~~~~~~~~~~~~
The best optimisation is always to just not do something. A ledger is meant to store relevant data and coordinate different 
parties. If you use the ledger as processing queue, you will run into performance issues, as each transaction view is cryptographically
secured and processed quite extensively to ensure privacy and integrity, which is totally unnecessary for intermediate steps. The rule 
of thumb is: if the output of a transaction submitted by a party is immediately consumed by the same party, then you are using the 
ledger as a processing queue. Instead, restructure your command such that the two steps happen as one.

Furthermore, each `create` will create a contract, causing data to be written to the ledger. Each `fetch` will cause the interpreter to halt 
interpretation, asking the ledger for a contract, which in worst case requires a lookup in the database. As the interpretation 

> must happen sequentially, this will mean a one by one lookup of contracts, causing load and latency. Fetching data is important and 

Reviewer: ```suggestion
must happen sequentially, contracts are looked up one by one, causing load and latency. Fetching data is important and 
```

Author: fixed

> must happen sequentially, this will mean a one by one lookup of contracts, causing load and latency. Fetching data is important and 

Author: fixed

a key feature, but you should apply the same reasoning as you would for a database: A database lookup is expensive and should only be 
done if necessary, ideally caching repetive computing results.

As an example, if you have a high throughput process that always resolves some data by a chain `A -> B -> C -> D` (four fetches), then 
change your model to use a single cache contract `ABCD`, only fetch that contract and ensure that there is a process that whenever one 
of the contracts changes, `ABCD` is updated.

Also, avoid unnecessary transient contracts, as they may cause additional views. `createAndExercise` is not a single command, 
but translated to one create and an exercise. Use a `nonconsuming` choice instead, possibly with a choice observer if you need 
to leave a trace on the ledger for audit purposes.

> Generally, using `lookupKey` is also discouraged. If you use `lookupByKey` for on-ledger deduplication, the lookup require a 

Reviewer: ```suggestion
Generally, using `lookupKey` is also discouraged. If you use `lookupByKey` for on-ledger deduplication, the lookup requires a 
```

Author: fixed

> database lookup, as in the absolut majority of the cases, the lookupByKey will resolve to `None`. Instead, use command deduplication 

Reviewer: ```suggestion
database lookup, as the lookupByKey will most likely resolve to `None`. Instead, use command deduplication 
```

Author: fixed

> database lookup, as in the absolut majority of the cases, the lookupByKey will resolve to `None`. Instead, use command deduplication 

Author: fixed
